streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java

```diff
@@ -157,26 +165,82 @@ public void shouldIdempotentlyUpdateSubscriptionFromActiveAssignment() {
+        );
+        final Map<TopicPartition, Long> task10ChangelogOffsets = mkMap(
+            mkEntry(t1p0, 20L),
+            mkEntry(t1p1, 30L)
+        );
+
+        final Long task00OffsetSum = Task.LATEST_OFFSET;
+        final Long task01OffsetSum = 3L;
+        final Long task02OffsetSum = 15L;
+        final Long task10OffsetSum = 50L;
 
-        assertThat((new File(taskFolders[0], StateManagerUtil.CHECKPOINT_FILE_NAME)).createNewFile(), is(true));
-        assertThat((new File(taskFolders[1], StateManagerUtil.CHECKPOINT_FILE_NAME)).createNewFile(), is(true));
-        assertThat((new File(taskFolders[3], StateManagerUtil.CHECKPOINT_FILE_NAME)).createNewFile(), is(true));
+        final Map<TaskId, Set<TopicPartition>> activeTaskAssignment = new HashMap<>(taskId00Assignment);
+        final StateMachineTask task00 = new StateMachineTask(taskId00, taskId00Partitions, true);
+        expectRestoreToBeCompleted(consumer, changeLogReader);
+        expect(activeTaskCreator.createTasks(anyObject(), eq(taskId00Assignment)))
+            .andReturn(singletonList(task00)).once();
+        final StateMachineTask task02 = new StateMachineTask(taskId02, taskId02Partitions, false);
+        task02.setChangelogOffsets(task02ChangelogOffsets);
+        expect(standbyTaskCreator.createTasks(eq(taskId02Assignment)))
+            .andReturn(singletonList(task02)).once();
+        final StateMachineTask task01 = new StateMachineTask(taskId01, taskId01Partitions, true);
+        task01.setChangelogOffsets(task01ChangelogOffsets);
+        expect(activeTaskCreator.createTasks(anyObject(), eq(taskId01Assignment)))
+            .andReturn(singletonList(task01)).once();
 
         expect(stateDirectory.listTaskDirectories()).andReturn(taskFolders).once();
 
-        replay(activeTaskCreator, stateDirectory);
+        expect(stateDirectory.lock(taskId10)).andReturn(true).once();
+        expect(stateDirectory.lock(taskId12)).andReturn(false).once();
+
+
+        final File task10CheckpointFile = new File(taskFolders[3], StateManagerUtil.CHECKPOINT_FILE_NAME);
+        final File task12CheckpointFile = new File(taskFolders[5], StateManagerUtil.CHECKPOINT_FILE_NAME);
+        assertThat(task10CheckpointFile.createNewFile(), is(true));
+        assertThat(task12CheckpointFile.createNewFile(), is(true));
+        new OffsetCheckpoint(task10CheckpointFile).write(task10ChangelogOffsets);
+
+        replay(activeTaskCreator, standbyTaskCreator, consumer, changeLogReader, stateDirectory);
+
+        taskManager.handleAssignment(activeTaskAssignment, taskId02Assignment);
+        assertThat(taskManager.tryToCompleteRestoration(), is(true));
+
+        activeTaskAssignment.putAll(taskId01Assignment);
+        taskManager.handleAssignment(activeTaskAssignment, taskId02Assignment);
+
+        assertThat(task00.state(), is(Task.State.RUNNING));
+        assertThat(task01.state(), not(Task.State.RUNNING));
+        assertThat(task02.state(), is(Task.State.RUNNING));
```

Comment:
I bet she copied the idiom from all of my tests. I did it because it makes the tests easier to read... I.e., you can visually see what state everything is in. Otherwise you'd have to reason about what state it _would_ be in, given all the mocks above.

Reply:
To verify that we're actually testing what we think we're testing, ie if due to some bug in `TaskManager` `0_0` did not actually reach the `RUNNING` state, we should fail fast. Otherwise when the test fails, it's not clear that it's due to an unrelated bug rather than a bug in `getTaskOffsetSums`

Reply:
Also what @vvcephei said 🙂 . It's kind of hard to reason about the states based only on calls to `TaskManager` in this test class

Reply:
Fair enough given the complexity of the setup. I guess what disturbs me most is the fact that the setup is so complex.